l3kernel/l3prop.dtx

```diff
@@ -49,32 +49,73 @@
 % entry overwrites the existing one. The \meta{keys} are compared on a
 % string basis, using the same method as \cs{str_if_eq:nnTF}.
 %
 % Property lists are intended for storing key-based information for use within
-% code.  This is in contrast to key--value lists, which are a form of
+% code. They can be converted from and to key--value lists, which are a form of
 % \emph{input} parsed by the \pkg{l3keys} module.
 %
+% Internally, property lists can use two distinct implementations with different
+% data storage, which are decided when declaring the property list variable
+% using \cs{prop_new:N} (\enquote{flat} storage) or \cs{prop_new_large:N}
+% (\enquote{linked} storage). Once a property list is declared with
+% \cs{prop_new:N} or \cs{prop_new_large:N}, the type of internal data storage
+% can no longer be changed. All other \pkg{l3prop} functions transparently
+% manipulate either storage method and convert as needed.
+% \begin{itemize}
+%   \item
+%     The (default) \enquote{flat} storage method is suited for a relatively
+%     small number of entries, or when the property list is likely to be
+%     manipulated (copied, mapped) as a whole rather than entry-wise.  It is
+%     significantly faster for \cs{prop_set_eq:NN}, and only slightly faster for
+%     \cs{prop_clear:N}, \cs{prop_concat:NNN}, and mapping functions
+%     \cs[no-index]{prop_map_\ldots{}}.
+%
+%   \item
+%     The \enquote{linked} storage method is meant for property lists with a
+%     large numbers of entries.  It has more memory overhead, but is
+%     significantly faster (for long lists) when accessing or modifying
+%     individual entries using functions such as \cs{prop_if_in:Nn},
+%     \cs{prop_item:Nn}, \cs{prop_put:Nnn}, \cs{prop_get:NnN},
+%     \cs{prop_pop:NnN}, \cs{prop_remove:Nn}, as it takes a constant
+%     time for these operations (rather than the number of items for a
+%     \enquote{flat} property list).  A technical drawback is that
+%     memory is permanently used\footnote{Until the end of the run, that
+%     is.} by \meta{keys} stored in a \enquote{linked} property list,
+%     even after they are removed and the property list is deleted.
+% \end{itemize}
+%
 % \section{Creating and initialising property lists}
 %
 % \begin{function}{\prop_new:N, \prop_new:c}
 %   \begin{syntax}
 %     \cs{prop_new:N} \meta{property list}
 %   \end{syntax}
-%   Creates a new \meta{property list} or raises an error if the name is
-%   already taken. The declaration is global. The \meta{property list}
-%   initially contains no entries.
+%   Creates a new \enquote{flat} \meta{property list} or raises an error if the
+%   name is already taken. The declaration is global. The \meta{property list}
+%   initially contains no entries.  See also \cs{prop_new_large:N}.
+% \end{function}
+%
+% \begin{function}[added = 2024-01-15]{\prop_new_large:N, \prop_new_large:c}
+%   \begin{syntax}
+%     \cs{prop_new_large:N} \meta{property list}
```

Comment:
You correctly pinpointed my reason to prefer "large" in function names (and my inconsistency in using a different word in error messages), which is that linked is an implementation detail.  In fact, the new props are both doubly-linked and implemented using the hash table, and the latter point is an important part of why they end up faster for accessing entries.  Is "linked" really a good word for that? "hashed"?

Reply:
I agree that names that refer to the implementation are not good (since there may be further changes to this in another decade or so). 

I also understand @FrankMittelbach ‘s problem with labelling them as “large”.

Needs further thought!

Reply:
Further: in some of the documentation, it is probably best to use both names.

Reply:
We need a name that does not imply any “ordering” on the type of implementation, and says nothing about the expected contents, but is also independent of any details of the implementation.